presto-main/src/main/java/io/prestosql/operator/scalar/time/TimeToTimestampCast.java

```diff
@@ -0,0 +1,96 @@
+import io.prestosql.spi.type.LongTimestamp;
+import io.prestosql.spi.type.TimeType;
+
+import java.time.Instant;
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.ZoneOffset;
+import java.util.List;
+
+import static io.prestosql.spi.function.OperatorType.CAST;
+import static io.prestosql.type.DateTimes.MICROSECONDS_PER_SECOND;
+import static io.prestosql.type.DateTimes.PICOSECONDS_PER_MICROSECOND;
+import static io.prestosql.type.DateTimes.PICOSECONDS_PER_SECOND;
+import static io.prestosql.type.DateTimes.SECONDS_PER_DAY;
+import static io.prestosql.type.DateTimes.rescale;
+import static io.prestosql.type.DateTimes.round;
+import static io.prestosql.type.DateTimes.scaleEpochMicrosToMillis;
+import static java.lang.Math.multiplyExact;
+
+@ScalarOperator(CAST)
+public final class TimeToTimestampCast
+{
+    private TimeToTimestampCast() {}
+
+    @LiteralParameters({"sourcePrecision", "targetPrecision"})
+    @SqlType("timestamp(targetPrecision)")
+    public static long castToShort(
+            @LiteralParameter("sourcePrecision") long sourcePrecision,
+            @LiteralParameter("targetPrecision") long targetPrecision,
+            ConnectorSession session,
+            @SqlType("time(sourcePrecision)") long time)
+    {
+        long epochMicros = computeEpochMicros(sourcePrecision, targetPrecision, session, time);
+        if (targetPrecision <= 3) {
+            return scaleEpochMicrosToMillis(epochMicros);
+        }
+
+        return epochMicros;
+    }
+
+    @LiteralParameters({"sourcePrecision", "targetPrecision"})
+    @SqlType("timestamp(targetPrecision)")
+    public static LongTimestamp castToLong(
+            @LiteralParameter("sourcePrecision") long sourcePrecision,
+            @LiteralParameter("targetPrecision") long targetPrecision,
+            ConnectorSession session,
+            @SqlType("time(sourcePrecision)") long time)
+    {
+        long epochMicros = computeEpochMicros(sourcePrecision, targetPrecision, session, time);
+        return new LongTimestamp(epochMicros, (int) (time % PICOSECONDS_PER_MICROSECOND));
+    }
+
+    private static long computeEpochMicros(@LiteralParameter("sourcePrecision") long sourcePrecision, @LiteralParameter("targetPrecision") long targetPrecision, ConnectorSession session, @SqlType("time(sourcePrecision)") long time)
+    {
+        long epochDay = LocalDate.ofInstant(session.getStart(), session.getTimeZoneKey().getZoneId())
+                .toEpochDay();
+
+        long epochSeconds = multiplyExact(epochDay, SECONDS_PER_DAY) + time / PICOSECONDS_PER_SECOND;
+
+        if (session.isLegacyTimestamp()) {
+            Instant instant = Instant.ofEpochSecond(epochSeconds);
+            LocalDateTime dateTime = LocalDateTime.ofInstant(instant, session.getTimeZoneKey().getZoneId());
+            List<ZoneOffset> offsets = session.getTimeZoneKey().getZoneId().getRules().getValidOffsets(dateTime);
+            epochSeconds -= offsets.get(0).getTotalSeconds();
```

Comment:
When there are 2 offsets the latter one usually wins (eg when parsing string), so `Iterables.getLast(offsets)`.

Also, add a comment why we do not need to handle the case when `offsets.isEmpty()`
Perhaps 
```
epochSeconds = ZonedDateTime.ofInstant(Instant.ofEpochSecond(epochSeconds), ZoneOffset.UTC)
                    .withZoneSameLocal(session.getTimeZoneKey().getZoneId())
                    .toInstant().getEpochSecond();
```

would avoid having a comment